### PR TITLE
BAEL-3338: A Guide to AuthenticationManagerResolver in Spring Security

### PR DESCRIPTION
**Pull request by maryarm**
Implements Basic Authentication flow with AuthenticationManagerResolver.

### CODE REVIEW

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverController.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+@RestController
+public class AuthResolverController {
+
+    @GetMapping("/customer/welcome")
+    public Mono<String> sayWelcomeToCustomer(Mono<Principal> principal) {
+        return principal
+                .map(Principal::getName)
+                .map(name -> String.format("Welcome to our site, %s!", name));
+    }
```
Comment (eric-martin) [inline spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverController.java]: When continuing a line, use a two-space indent

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverController.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    @GetMapping("/employee/welcome")
+    public Mono<String> sayWelcomeToEmployee(Mono<Principal> principal) {
+        return principal
+                .map(Principal::getName)
+                .map(name -> String.format("Welcome to our company, %s!", name));
```
Comment (eric-martin) [inline spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverController.java]: When continuing a line, use a two-space indent

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        return filter;
+    }
+
+    public ReactiveAuthenticationManagerResolver<ServerHttpRequest> authenticationManagerResolver() {
+        return request -> {
+            if (request
+              .getPath()
+              .subPath(0)
+              .value()
+              .startsWith("/employee")) return Mono.just(employeesAuthenticationManager());
```
Comment (eric-martin) [inline spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java]: put the return on the next line with a two-space indent

File: spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    public AuthenticationConverter authenticationConverter() {
+        return new BasicAuthenticationConverter();
+    }
+
+    public AuthenticationManagerResolver<HttpServletRequest> authenticationManagerResolver() {
+        return request -> {
+            if (request
+              .getPathInfo()
+              .startsWith("/employee")) return employeesAuthenticationManager();
```
Comment (eric-martin) [inline spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java]: return on the next line

File: spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        AuthenticationFilter ret = new AuthenticationFilter(resolver, converter);
+        ret.setSuccessHandler((httpServletRequest, httpServletResponse, authentication) -> {
+        });
+        return ret;
+    }
+
+    public AuthenticationManager employeesAuthenticationManager() {
+        return authentication -> {
+            if (isEmployee(authentication)) {
+                return new UsernamePasswordAuthenticationToken(authentication.getPrincipal(), authentication.getCredentials(), Arrays.asList(new SimpleGrantedAuthority("ROLE_USER")));
```
Comment (eric-martin) [inline spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java]: break up this long line

File: spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+              .getPathInfo()
+              .startsWith("/employee")) return employeesAuthenticationManager();
+            return customersAuthenticationManager();
+        };
+    }
+
+    public AuthenticationManager customersAuthenticationManager() {
+        return authentication -> {
+            if (isCustomer(authentication)) {
+                return new UsernamePasswordAuthenticationToken(authentication.getPrincipal(), authentication.getCredentials(), Arrays.asList(new SimpleGrantedAuthority("ROLE_USER")));
```
Comment (eric-martin) [inline spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java]: break up this long line

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            return Mono.just(customersAuthenticationManager());
+        };
+    }
+
+    public ReactiveAuthenticationManager customersAuthenticationManager() {
+        return authentication -> customer(authentication)
+          .switchIfEmpty(Mono.error(new UsernameNotFoundException(authentication
+            .getPrincipal()
+            .toString())))
+          .map(b -> new UsernamePasswordAuthenticationToken(authentication.getPrincipal(), authentication.getCredentials(), Collections.singletonList(new SimpleGrantedAuthority("ROLE_USER"))));
```
Comment (eric-martin) [inline spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java]: break up this long line

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    public ReactiveAuthenticationManager employeesAuthenticationManager() {
+        return authentication -> employee(authentication)
+          .switchIfEmpty(Mono.error(new UsernameNotFoundException(authentication
+            .getPrincipal()
+            .toString())))
+          .map(b -> new UsernamePasswordAuthenticationToken(authentication.getPrincipal(), authentication.getCredentials(), Collections.singletonList(new SimpleGrantedAuthority("ROLE_USER"))));
```
Comment (eric-martin) [inline spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java]: break up this long line

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java (hunk 3 above)
Comment (maryarm) [reply spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java]: Fixed

File: spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java (hunk 4 above)
Comment (maryarm) [reply spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java]: Fixed

File: spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java (hunk 5 above)
Comment (maryarm) [reply spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java]: Fixed

File: spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java (hunk 6 above)
Comment (maryarm) [reply spring-5-security/src/main/java/com/baeldung/authresolver/AuthResolverWebSecurityConfigurer.java]: Fixed

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java (hunk 7 above)
Comment (maryarm) [reply spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java]: Fixed

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java (hunk 8 above)
Comment (maryarm) [reply spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/AuthResolverSecurityConfig.java]: Fixed

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/CustomWebSecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    public ReactiveAuthenticationManagerResolver<ServerHttpRequest> resolver() {
+        return request -> {
+            if (request
+              .getPath()
+              .subPath(0)
+              .value()
+              .startsWith("/employee"))
+                return Mono.just(employeesAuthenticationManager());
```
Comment (KevinGilmore) [inline spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/CustomWebSecurityConfig.java]: please enclose the single statement of the if clause in curly braces as per recommended practice

File: spring-5-security/src/main/java/com/baeldung/authresolver/CustomWebSecurityConfigurer.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    public AuthenticationConverter authenticationConverter() {
+        return new BasicAuthenticationConverter();
+    }
+
+    public AuthenticationManagerResolver<HttpServletRequest> resolver() {
+        return request -> {
+            if (request
+              .getPathInfo()
+              .startsWith("/employee"))
+                return employeesAuthenticationManager();
```
Comment (KevinGilmore) [inline spring-5-security/src/main/java/com/baeldung/authresolver/CustomWebSecurityConfigurer.java]: add curly braces

Comment (maryarm) [reply spring-5-security/src/main/java/com/baeldung/authresolver/CustomWebSecurityConfigurer.java]: -Fixed

File: spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/CustomWebSecurityConfig.java (hunk 9 above)
Comment (maryarm) [reply spring-5-reactive-security/src/main/java/com/baeldung/reactive/authresolver/CustomWebSecurityConfig.java]: -Fixed

**Pull request merged**